src/my-bot.ts

```diff
@@ -1,38 +1,423 @@
+                continue;
+            }
+            shipPlaces[3] = this.getShipPlace(shipPlaces,3);
+            if (shipPlaces[3].StartingSquare.Column === -1) {
+                continue;
+            }
+            shipPlaces[4] = this.getShipPlace(shipPlaces,2);
+            if (shipPlaces[4].StartingSquare.Column === -1) {
+                continue;
+            }
+            done = true;
+        } while ((!done) && (counter < 10000))
+        if (counter === 10000) {
+            throw new Error("Infinite loop generating ship positions");
         }
-        return { Row: "A", Column: 1 };  
+        return shipPlaces;
     }
 
-    private getNextTarget(position) {
+    public selectTarget(gamestate): Promise<Position> {
+        return new Promise((resolve, reject) => {
+            firebase.database().ref('matches/' + this.matchId.toString()).once('value').then((snapshot) => { 
+                let snapCopy = snapshot.val();
+                if (gamestate.MyShots.length === 0) {
+                    resolve(this.randomShot(gamestate));
+                }
+                const previousShot: Shot = gamestate.MyShots && gamestate.MyShots[gamestate.MyShots.length-1];
+                let target: Position = { Row: 'A', Column: 1 };
+                if (previousShot.WasHit) {
+                    snapCopy.hitmode = true;
+                    if (snapCopy.hitmap) {
+                        snapCopy.hitmap.push(previousShot.Position);
+                    } else {
+                        snapCopy.hitmap = [previousShot.Position];
+                    }
+                }
+                firebase.database().ref('matches/' + this.matchId.toString()).set(snapCopy);
+                if (snapCopy.hitmode) {
+                    target = this.track(gamestate,snapCopy);
+                } else if (snapCopy.hitmap) {
+                    target = this.randomShot(gamestate,snapCopy.hitmap);
+                } else {
+                    target = this.randomShot(gamestate);
+                }
+                return resolve(target);
+            })
+            .catch((err) => {console.log(err)});
+        })
+        
+        
+    }
+
+    private getNextTarget(position: Position): Position {
         var column = this.getNextColumn(position.Column);
         var row = column === 1 ? this.getNextRow(position.Row) : position.Row;
         return { Row: row, Column: column }
     }
 
-    private getNextRow(row) {
+    private getNextRow(row): string {
         var newRow = row.charCodeAt(0) + 1;
         if(newRow > 'J'.charCodeAt(0)) {
             return 'A';
         }
         return String.fromCharCode(newRow);
     }
 
-    private getNextColumn(column) {
+    private getNextColumn(column): number {
         return column % 10 + 1;
     }
+
+    private authenticate(): void {
+        firebase.auth().signInWithEmailAndPassword(
+            "david.may-miller@softwire.com",
+            "securePassword3"

```

Comment:
Woop.

(Obviously, in general, hardcoding passwords in git history is a bad thing, but it's very practical (the DMP project also has this atm)